### ISSUE TITLE
Question Regarding Caching in 4DMOS

### ISSUE THREAD

**Issue by MinChoi0129** [PRBonn/4DMOS#56]
Hello, thank you for providing 4DMOS.

I've observed that 4DMOS involves a data **pre-caching** process during execution. When using the **SemanticKITTI** dataset, this process works seamlessly as it's a single dataset. However, when using the **HeLiMOS** dataset, I'm wondering if the caching strategy needs to be adjusted based on the specific **training scenario**.

Specifically, when utilizing the **`all_training.yaml`** configuration, the cached data likely includes information from Avia, Aeva, Velodyne, and Ouster sensors. My question is: if I intend to train with **`omni_training`** later, will I need to **re-cache** the data entirely? Or will the system automatically extract and use only the relevant "omni" portion from the cache generated during the `all_training` process?

Your clarification on this matter would be greatly appreciated.

----
Questions are translated using Gemini.

**Comment by benemer** [comment]
Yes, you can cache all the data using the `all_training.yaml` config and then use different subsets for training, for example `omni_training.yaml`. We cache the output of the function `get_past_point_clouds` in the `mos4d_dataset.py`:

```python
    @memoize()
    def get_past_point_clouds(
        self,
        sequence: int,
        scan_index: int,
        n_past_scans: int,
        data_config_dict: Dict,
        odometry_config_dict: Dict,
    ):
```
 
If your training config only differs in the sequences and all training sequences have been cached, there is no need to re-cache. However, if you want to use a different number of past scans or use a different config for the data or odometry, this will trigger a re-caching (which must not happen during training). This is because there is no cached output of the `get_past_point_clouds` with these arguments.

You can always verify by checking if there is a print `*****Caching now*****`. This should never happen during training.

**Comment by MinChoi0129** [comment]
Cool! Thank you for answering.

I've noticed that you are the second author of the HeLiMOS paper, and I have a few questions regarding both the HeLiMOS paper and the 4DMOS repository.

First, regarding the evaluation in the HeLiMOS paper:

1.  The HeLiMOS dataset is divided into train, validation, and test sets. Are the Mean IoU scores reported in **Table II and Table III** of the HeLiMOS paper based on the **validation set or the test set**? Additionally, for the 4DMOS evaluation, are these scores from the **'online' or 'receding'** setting?

2.  The primary metric in the HeLiMOS paper is **Mean IoU**. I am curious why Mean IoU was chosen over a standard IoU. Specifically, what is being averaged to calculate the "Mean" in Mean IoU? (To clarify, I understand the "Avg" and "Total" rows in the tables; my question is about the "Mean" component of the metric itself).

3.  In the 4DMOS repository's evaluation process for HeLiMOS, my understanding is that the validation loader processes all ~13,000 frames for inference, and then evaluation is performed only on the frames that have corresponding labels. Is this understanding correct? If so, could you explain the reasoning for performing inference on the entire sequence, rather than constructing the validation loader using only the labeled frames specified in `val.txt`?

4.  I would like to know the **hyperparameters** (e.g., epochs, batch size, etc.) used for training and evaluating 4DMOS and MapMOS on the HeLiMOS dataset. Were the parameters identical to those used for the original datasets in the 4DMOS and MapMOS papers?

Thank you for your time and assistance.

**Comment by benemer** [comment]
1. The numbers are reported for the test set. We only use the validation set for deciding which trained model to use (based on the highest IoU). For 4DMOS, we call the receding horizon strategy `delayed` to make it clearer that it introduces a delay. In Table III, we report the results of 4DMOS with delay and for MapMOS with the volumetric belief since they delivered the best results in Table II.

2. We compute the IoU for each sensor by aggregating the confusion matrix for all scans in the set and then computing the IoU with Equation 2. For the two other columns, we report the mean of per-sensor IoUs, which is why it is called mean IoU. Another option would be to always compute an IoU by getting the aggregated confusion matrix for each setup, which will lead to different results.

3. For inference, we always process the whole sequence and then only evaluate the results for available labels. This, of course, slows down the evaluation, because some clouds are segmented but not evaluated. If you only infer the labels for scans with ground truth data, the prediction for a specific scan can be different depending on which labels you evaluate, since some scans are skipped for inference and therefore not part of what the model sees. Short example: Let's say you have 10 scans and labels for the last 5. We infer the labels for all 10 scans independently of the availability of ground truth and then evaluate the last 5 predictions. If you instead skip the first 5 frames and start with the last 5, the prediction for frame 5 is already different (since now it is the first in the sequence, whereas before it was right in the middle). This is, of course, also possible, but requires more modification of the inference code (what if every second scan or some keyframes are labeled?). For example, nuScenes only has annotations for some keyframes, and only passing those labeled scans to the model will lead to poor results due to the time gap between them.

4. We provide all configurations used in the paper under `config/helimos`. We use the same configurations for all sensors as in the original publications and only change the name of the sequences.
